File: aicode/minicli2/src/tools/grep.py
"""Code search tool."""

import re
from pathlib import Path
from typing import Optional
from dataclasses import dataclass


@dataclass
class GrepResult:
    """Grep result."""

    success: bool
    matches: list[dict] = None
    error: str = ""

    def __post_init__(self):
        if self.matches is None:
            self.matches = []


class GrepTool:
    """Search code files."""

    def __init__(self, allowed_dirs: Optional[list[Path]] = None):
        """Initialize grep tool.

        Args:
            allowed_dirs: List of allowed directories. Defaults to current directory.
        """
        self.allowed_dirs = allowed_dirs or [Path.cwd()]

    def _is_path_allowed(self, path: Path) -> bool:
        """Check if path is within allowed directories.

        Args:
            path: Path to check.

        Returns:
            True if allowed, False otherwise.
        """
        try:
            resolved = path.resolve()
            for allowed in self.allowed_dirs:
                allowed_resolved = allowed.resolve()
                if resolved.is_relative_to(allowed_resolved):
                    return True
            return False
        except (ValueError, OSError):
            return False

    def search(
        self,
        pattern: str,
        path: str = ".",
        file_pattern: str = "*",
        context_lines: int = 2,
    ) -> GrepResult:
        """Search for pattern in files.

        Args:
            pattern: Search pattern (regex supported).
            path: Search path.
            file_pattern: File pattern (glob).
            context_lines: Number of context lines.

        Returns:
            Grep result.
        """
        try:
            search_path = Path(path).expanduser()

            if not self._is_path_allowed(search_path):
                return GrepResult(
                    success=False,
                    error=f"Access denied: {path} is not within allowed directories",
                )

            if not search_path.exists():
                return GrepResult(
                    success=False,
                    error=f"Path not found: {path}",
                )

            # Compile regex
            try:
                regex = re.compile(pattern)
            except re.error as e:
                return GrepResult(
                    success=False,
                    error=f"Invalid regex pattern: {e}",
                )

            matches = []
            files = search_path.rglob(file_pattern)

            for file_path in files:
                if not file_path.is_file():
                    continue

                if not self._is_path_allowed(file_path):
                    continue

                # Skip binary files
                try:
                    with open(file_path, "r", encoding="utf-8") as f:
                        lines = f.readlines()
                except (UnicodeDecodeError, PermissionError):
                    continue

                for line_num, line in enumerate(lines, start=1):
                    if regex.search(line):
                        matches.append({
                            "file": str(file_path),
                            "line": line_num,
                            "content": line.rstrip(),
                        })

            return GrepResult(success=True, matches=matches)

        except Exception as e:
            return GrepResult(success=False, error=str(e))

    @property
    def definition(self) -> dict:
        """Get tool definition for API.

        Returns:
            Tool definition dict.
        """
        return {
            "type": "function",
            "function": {
                "name": "grep",
                "description": "Search for text patterns in code files using regex. Returns matching lines with file paths and line numbers.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "pattern": {
                            "type": "string",
                            "description": "The regex pattern to search for",
                        },
                        "path": {
                            "type": "string",
                            "description": "The directory path to search in (default: current directory)",
                        },
                        "file_pattern": {
                            "type": "string",
                            "description": "File glob pattern to filter files (e.g., '*.py', '*.js')",
                        },
                    },
                    "required": ["pattern"],
                },
            },
        }


def grep(pattern: str, path: str = ".", file_pattern: str = "*") -> str:
    """Search for pattern and return results as string.

    Args:
        pattern: Search pattern.
        path: Search path.
        file_pattern: File pattern.

    Returns:
        Search results or error message.
    """
    tool = GrepTool()
    result = tool.search(pattern, path, file_pattern)

    if not result.success:
        return f"Error: {result.error}"

    if not result.matches:
        return "No matches found."

    output_lines = []
    for match in result.matches:
        output_lines.append(f"{match['file']}:{match['line']}: {match['content']}")

    return "\n".join(output_lines)

File: aicode/minicli6/src/tools/impl/Grep.ts
import { Tool, ToolCall, ToolResult, ToolContext } from '../Tool.js';
import { readdir, stat, readFile } from 'fs/promises';
import path from 'path';

interface MatchResult {
  file: string;
  line: number;
  content: string;
}

export class GrepTool extends Tool {
  name = 'grep';
  description = 'Fast content search tool that works with any codebase size. Searches file contents using regular expressions. Supports full regex syntax (e.g., "log.*Error", "function\\s+\\w+"). Returns file paths and line numbers with matches, sorted by modification time.';
  parameters = {
    type: 'object' as const,
    properties: {
      pattern: {
        type: 'string' as const,
        description: 'Regular expression pattern to search for in file contents',
      },
      path: {
        type: 'string' as const,
        description: 'Directory to search in (default: current working directory)',
      },
      include: {
        type: 'string' as const,
        description: 'File pattern to include (e.g., *.ts, *.{js,ts})',
      },
    },
    required: ['pattern'],
  };
  permission: 'auto' = 'auto';

  private async walkDir(dir: string, results: string[]): Promise<void> {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!['node_modules', '.git', 'dist', 'build', '.next'].includes(entry.name)) {
          await this.walkDir(fullPath, results);
        }
      } else if (entry.isFile()) {
        results.push(fullPath);
      }
    }
  }

  private matchesPattern(filename: string, includePattern?: string): boolean {
    if (!includePattern) return true;

    const patterns = includePattern.replace(/^\*\./, '').split(',').map(p => p.trim().replace(/^\*\.?/, ''));
    const ext = path.extname(filename).slice(1);
    return patterns.some(p => ext === p || filename.endsWith(p));
  }

  async execute(call: ToolCall, context: ToolContext): Promise<ToolResult> {
    const { pattern, path: searchPath = context.workingDirectory, include } = call.arguments;

    let regex: RegExp;
    try {
      regex = new RegExp(pattern, 'gm');
    } catch (error) {
      return {
        success: false,
        output: '',
        error: `Invalid regex pattern: ${error instanceof Error ? error.message : String(error)}`,
      };
    }

    let files: string[];
    try {
      files = [];
      await this.walkDir(searchPath, files);
    } catch (error) {
      return {
        success: false,
        output: '',
        error: `Failed to read directory: ${error instanceof Error ? error.message : String(error)}`,
      };
    }

    const filteredFiles = files.filter(f => this.matchesPattern(f, include));

    const matches: MatchResult[] = [];
    const fileStats: Map<string, number> = new Map();

    await Promise.all(
      filteredFiles.map(async (file) => {
        try {
          const [content, stats] = await Promise.all([
            readFile(file, 'utf-8'),
            stat(file),
          ]);
          
          fileStats.set(file, stats.mtime.getTime());
          
          const lines = content.split('\n');
          for (let i = 0; i < lines.length; i++) {
            regex.lastIndex = 0;
            if (regex.test(lines[i])) {
              matches.push({
                file,
                line: i + 1,
                content: lines[i].trim().slice(0, 200),
              });
            }
          }
        } catch {
          // Skip files that can't be read
        }
      })
    );

    matches.sort((a, b) => (fileStats.get(b.file) || 0) - (fileStats.get(a.file) || 0));

    const uniqueFiles = [...new Set(matches.map(m => m.file))];

    if (matches.length === 0) {
      return {
        success: true,
        output: `No matches found for pattern "${pattern}" in ${searchPath}`,
      };
    }

    const output = matches
      .slice(0, 100)
      .map(m => `${m.file}:${m.line}: ${m.content}`)
      .join('\n');

    return {
      success: true,
      output: output + (matches.length > 100 ? `\n... and ${matches.length - 100} more matches` : ''),
      metadata: {
        totalMatches: matches.length,
        filesWithMatches: uniqueFiles.length,
        pattern,
        path: searchPath,
      },
    };
  }
}
